### ISSUE TITLE
Potential bug of MoE training with zero3

### ISSUE THREAD

**Issue by Opdoop** [ScienceOne-AI/DeepSeek-671B-SFT-Guide#7]
During training, some experts don't see any tokens; thus, there are no gradients, and all other processes will get stuck.
https://github.com/deepspeedai/DeepSpeed/issues/5066

Known Issues:
When training the Dense+MoE hybrid model, the process will hang if any PP rank does not have expert params.
https://github.com/NVIDIA/Megatron-LM/releases/tag/v0.11.0

**Comment by Opdoop** [comment]
Maybe it is the reason why NCCL times out after some training steps:
![Image](https://github.com/user-attachments/assets/6bb0f7a3-a317-4999-8a77-8d3edd5adda8)
